pwmig/exportgrid/exportgrid.1
.TH EXPORTGRID 1 "$Date: 2009/06/19 12:30:38 $"
.SH NAME
exportgrid - ascii output of GCLgrid objects
.SH SYNOPSIS
.nf
exportgrid db|fname gridname [-fin -f fieldname -3d -vector n -o outfile -dir outdir] 
.fi
.SH DESCRIPTION
.LP
This program exports various 
GCLgrid related objects as simple ascii data.  This implementation supports
6 data type:  GCLgrid, GCLgrid3d, GCLscalarfield, GCLscalarfield3d,
GCLvectorfield, and GCLvectorfield3d.  
The program will always write results to stdout.  It will optionally operate 
in what is effectively a conversion mode.   That is, the -o flag can be used
to export data from the database to a file format that can be read by some program
that use the GCLgrid library.   The stdout output, in contrast, is always plain
ascii coded data.
.LP
The gclgrid library has two ways of storing data:  (1) in a database, and (2) in 
a mixed binary/pf format.   The default assumes the data are in a datascope 
database defined by the db argument.  
If the -fin flag appears the program assumes that argv[1]
is not a database but a root file name used for the file based output of a gclgrid.
The file based input assumes two files are present with the root name defined by the fname
argument (e.g. if fname is "migdata1" the program will look for binary data in "migdata.dat"
and related parametric data is stored in "migdata.pf", which is cracked with standard pf 
routines.)  
.SH OPTIONS
.IP -f
Dump the field with the unique name \fIfieldname\fR defined in the gclfield
table.  The field and grid names must be consistent or the program will
abort.
.IP -3d
Switch to 3d mode.  The default assumes the grid/field are 2d objects.
.IP -vector
Assume the field being referenced is a vector field object.  
The argument following is the expected vector size.   An error is thrown if this does 
not match the size stored with the data.   
.IP -o
If this flag appears the program will write to the name defined by outfile. 
This is needed for the conversion mode noted the DESCRIPTION section above.
.IP -dir
When -dir appears the output will be written to a directory with the name dir.
This flag is ignored unless the -o flag appears. If -o is used without an
associated -dir pair output will be written in the current directory.  
.LP
The -f, -3d, and -vector options work together to define the type of 
object the program tries to load.  If -f is not used the assumption is
only grid geometry is to be written to output.  The -3d option is a
switch between the class of 2d and 3d GCLgrid/field objects. Finally,
the -vector argument is required if a field being referenced is a 
vector field. 
.SH FILES
.ce
\fISTDOUT text files\fR
.LP
At this time the output format is very different for field objects
and grid objects.  The format for grid object is simpler and is just a series
of blank separated fields defining (in order):  latitude (degrees),
longitude (degrees), and depth (km).  
.LP
For field data this program is little more than a front end for
the C++ operator << for these objects.  The format is:
.LP
\fILine 1:\fR  is a blank separated list of two or three integers; 
two for 2d and three for 3d data.  These define the number of node
points in x1, x2, and x3 directions respectively.  
.LP
The remainder of the file is a series of lines with a count equal to 
the product of the integers printed on line 1.  Each line of these
lines is a blank separated list of attributes in the following order:
x1, x2, x3, latitude, longitude, radius, val(1), ..., val(nv)
where x1,x2, and x3 are the location of this grid point in the 
Cartesian system used internally by the GCLgrid library; 
latitude, longitude, and radius are Earth coordinates, and 
val(1) to val(nv) are the nv data values associated with this 
grid point.  For scalar fields nv is 1 so there is only one val.  
.ce
\fIGrid File Output\fR
.LP
The format used to store a GCLgrid with the -o option is a version of
the two file solution for data and metadata.  Two files are created:  one
with a ".dat" appendage and one with a ".pf" extension.   The first is 
binary dump of the coordinates of the grid followed (for formats) with a 
dump the data attached to each grid point.   The pf stores the attributes
that describe the contents of the .dat file.   They are identical to 
attribute names in the gclgdisk and gclfield tables used for database
indexing.
.SH "SEE ALSO"
.nf
http://geology.indiana.edu/pavlis/software/gclgrid/html/annotated.html
.fi
.SH ERRORS
.LP
Be aware it is always a fatal error to mix -fin and -o.   The -o option 
is intended strictly as a converter.
.SH "BUGS AND CAVEATS"
.IP (1)
The program does not automatically make a sanity check for scalar versus
vector fields.  The output will be scrambled if a vector field is referenced
without using the -vector flag.  I am not sure what will happen if a 
scalar field is referenced with the -vector flag set.  
Remember that famous quote:  "Do you feel lucky?  Do ya, punk?".
.IP (2)
There probably should be the reverse converter to import file data into a database.
The conversion mode was an add on and the clumsy usage shows.
.IP (3) 
Requiring the user to specify the vector size is an unnecessary complexity but 
one I found useful as a sanity check. 
.SH AUTHOR
.nf
Gary L. Pavlis
Indiana University
pavlis@indiana.edu
.fi
.\" $Id: exportgrid.1,v 1.2 2009/06/19 12:30:38 pavlis Exp $


pwmig/exportgrid/exportgrid.cc
#include <stdio.h>
#include "stock.h"
#include "seispp.h"
#include "dbpp.h"
#include "gclgrid.h"
using namespace std;
using namespace SEISPP;
void usage()
{
	cerr << "exportgrid db|fname gridname [-fin -f fieldname -3d -vector -o outfile -dir outdir] "<<endl
		<< " outputs GCLgrid or field objects to stdout"<<endl
                << " -fin selects file based input with name fname (arg1) - default is db input"<<endl
		<< " Dumps only grid unless -f is used."<<endl
                << " Use -3d if data are defined on a 3d grid (default is 2d)"<<endl
                << " Use -vector if the data are a vector field (default is scalar)"<<endl
                << " -o saves to outfile using pfhdr format (used mainly for conversion from db)"<<endl
                << " -dir defines directory where file based saves will be done (default is curent directory)"<<endl
                << "  ( -o not allowed with -fin option)"<<endl;
	exit(-1);
}
bool SEISPP::SEISPP_verbose(true);
int main(int argc, char **argv)
{
	elog_init(argc,argv);
        /* This is defined in gclgrid.h and perhaps should use the symbol default_output_format
           used there, but we will code it here.  Potential maintenance issue. */
        const string formstr("pfhdr");
	if(argc < 3) usage();
	string dbname(argv[1]);
        string fname;
	string gridname(argv[2]);
	string fieldname("");  //initialization for safety
        bool dbmode(true);
	bool threedmode(false);
	bool isfield(false);
	bool vectorfield(false);
        bool write_file_output(false);
        string outfile;
        string outdir(".");
	for(int i=3;i<argc;i++)
	{
		string sarg(argv[i]);
		if(sarg=="-3d") 
			threedmode=true;
                else if(sarg=="-fin")
                {
                    dbmode=false;
                    fname=dbname;
                }
		else if(sarg=="-f")
		{
			++i;
			if(i==argc) usage();
                        write_file_output=true;
			fieldname=string(argv[i]);
			isfield=true;
		}
		else if(sarg=="-vector")
			vectorfield=true;
                else if(sarg=="-dir")
                {
			++i;
			if(i==argc) usage();
                        outdir=string(argv[i]);
                }
                else if(sarg=="-o")
                {
			++i;
			if(i==argc) usage();
                        outfile=string(argv[i]);
                }
		else
			usage();
	}
        if(write_file_output && (!dbmode)) usage();
	try {
		DatascopeHandle dbh;
                if(dbmode) dbh=DatascopeHandle(dbname,true);
		if(threedmode)
		{
		    if(isfield)
		    {
			if(vectorfield)
			{
                            if(dbmode)
                            {
				GCLvectorfield3d field(dbh,gridname,fieldname);
				cout << field;
                                if(write_file_output) field.save(outfile,outdir);
                            }
                            else
                            {
                                GCLvectorfield3d field(fname,formstr);
                                cout << field;
                            }
			}
			else
			{
                            if(dbmode)
                            {
				GCLscalarfield3d field(dbh,gridname,fieldname);
				cout << field;
                                if(write_file_output) field.save(outfile,outdir);
                            }
                            else
                            {
                                GCLscalarfield3d field(fname,formstr);
                                cout << field;
                            }
			}
		     }
		     else
		     {
			GCLgrid3d grid;
                        if(dbmode)
                            grid=GCLgrid3d(dbh,gridname);
                        else
                            grid=GCLgrid3d(fname,formstr);
			/* fields define operator <<, but there is this is not
			currently defined for grids.  We write a grid in a simpler
			format*/
			for(int k=0;k<grid.n3;++k)
			  for(int j=0;j<grid.n2;++j)
			    for(int i=0;i<grid.n1;++i)
				cout << deg(grid.lon(i,j,k)) <<" "
					<<deg(grid.lat(i,j,k))<<" "
					<<grid.depth(i,j,k)<<endl;
                         if(write_file_output) grid.save(outfile,outdir);
			
		     }
		}
		else
		{
		    if(isfield)
		    {
			if(vectorfield)
			{
				GCLvectorfield field(dbh,gridname,fieldname);
				cout << field;
                                if(write_file_output) field.save(outfile,outdir);
			}
			else
			{
				GCLscalarfield field(dbh,gridname,fieldname);
				cout << field;
                                if(write_file_output) field.save(outfile,outdir);
			}
		     }
		     else
		     {
			GCLgrid grid;
                        if(dbmode)
                            grid=GCLgrid(dbh,gridname);
                        else
                            grid=GCLgrid(fname,formstr);
			for(int j=0;j<grid.n2;++j)
			    for(int i=0;i<grid.n1;++i)
				cout << deg(grid.lon(i,j)) <<" "<<deg(grid.lat(i,j))<<endl;
                        if(write_file_output) grid.save(outfile,outdir);
		     }
		}
	} catch (int ierr)
	{
		cerr << "The GCL object constructor called threw exception ="<<ierr<<endl;
		cerr << "Antelope elog message:"<<endl;
		elog_die(0,"exportgrid:  failed");
	}
	catch (SeisppError& serr)
	{
		serr.log_error();
		usage();
	}
        catch (std::exception& stexp)
        {
            cerr << stexp.what()<<endl;
        }
}
